File: python/design_pattern/visitor/main_visitor.py
# /// script
# requires-python = ">=3.10"
# dependencies = []
# ///
"""
Visitor Pattern

The Visitor pattern lets you define a new operation without changing the
classes of the elements on which it operates. It separates algorithms from
the object structure they operate on.

Key Components:
- Visitor: Declares visit operations for each ConcreteElement
- ConcreteVisitor: Implements operations for each element type
- Element: Declares accept method that takes a Visitor
- ConcreteElement: Implements accept to call appropriate visit method
- ObjectStructure: Collection of elements that can be iterated
"""

from abc import ABC, abstractmethod
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    pass


# Visitor interface
class ShapeVisitor(ABC):
    """Abstract visitor for shape operations."""

    @abstractmethod
    def visit_circle(self, circle: "Circle") -> str:
        pass

    @abstractmethod
    def visit_rectangle(self, rectangle: "Rectangle") -> str:
        pass

    @abstractmethod
    def visit_triangle(self, triangle: "Triangle") -> str:
        pass

    @abstractmethod
    def visit_compound(self, compound: "CompoundShape") -> str:
        pass


# Element interface
class Shape(ABC):
    """Abstract element that can accept a visitor."""

    @abstractmethod
    def accept(self, visitor: ShapeVisitor) -> str:
        pass


# Concrete Elements
class Circle(Shape):
    """Concrete element: Circle."""

    def __init__(self, x: float, y: float, radius: float):
        self.x = x
        self.y = y
        self.radius = radius

    def accept(self, visitor: ShapeVisitor) -> str:
        return visitor.visit_circle(self)


class Rectangle(Shape):
    """Concrete element: Rectangle."""

    def __init__(self, x: float, y: float, width: float, height: float):
        self.x = x
        self.y = y
        self.width = width
        self.height = height

    def accept(self, visitor: ShapeVisitor) -> str:
        return visitor.visit_rectangle(self)


class Triangle(Shape):
    """Concrete element: Triangle."""

    def __init__(
        self, x1: float, y1: float, x2: float, y2: float, x3: float, y3: float
    ):
        self.x1, self.y1 = x1, y1
        self.x2, self.y2 = x2, y2
        self.x3, self.y3 = x3, y3

    def accept(self, visitor: ShapeVisitor) -> str:
        return visitor.visit_triangle(self)


class CompoundShape(Shape):
    """Composite element containing multiple shapes."""

    def __init__(self, name: str):
        self.name = name
        self._shapes: list[Shape] = []

    def add(self, shape: Shape) -> None:
        self._shapes.append(shape)

    def get_shapes(self) -> list[Shape]:
        return self._shapes.copy()

    def accept(self, visitor: ShapeVisitor) -> str:
        return visitor.visit_compound(self)


# Concrete Visitors
class AreaCalculator(ShapeVisitor):
    """Visitor that calculates area of shapes."""

    def visit_circle(self, circle: Circle) -> str:
        import math

        area = math.pi * circle.radius**2
        return f"Circle area: {area:.2f}"

    def visit_rectangle(self, rectangle: Rectangle) -> str:
        area = rectangle.width * rectangle.height
        return f"Rectangle area: {area:.2f}"

    def visit_triangle(self, triangle: Triangle) -> str:
        # Using shoelace formula
        area = abs(
            (triangle.x1 * (triangle.y2 - triangle.y3))
            + (triangle.x2 * (triangle.y3 - triangle.y1))
            + (triangle.x3 * (triangle.y1 - triangle.y2))
        ) / 2
        return f"Triangle area: {area:.2f}"

    def visit_compound(self, compound: CompoundShape) -> str:
        total = 0.0
        for shape in compound.get_shapes():
            result = shape.accept(self)
            # Extract number from result string
            value = float(result.split(": ")[1])
            total += value
        return f"Compound '{compound.name}' total area: {total:.2f}"


class PerimeterCalculator(ShapeVisitor):
    """Visitor that calculates perimeter of shapes."""

    def visit_circle(self, circle: Circle) -> str:
        import math

        perimeter = 2 * math.pi * circle.radius
        return f"Circle perimeter: {perimeter:.2f}"

    def visit_rectangle(self, rectangle: Rectangle) -> str:
        perimeter = 2 * (rectangle.width + rectangle.height)
        return f"Rectangle perimeter: {perimeter:.2f}"

    def visit_triangle(self, triangle: Triangle) -> str:
        import math

        side1 = math.sqrt(
            (triangle.x2 - triangle.x1) ** 2 + (triangle.y2 - triangle.y1) ** 2
        )
        side2 = math.sqrt(
            (triangle.x3 - triangle.x2) ** 2 + (triangle.y3 - triangle.y2) ** 2
        )
        side3 = math.sqrt(
            (triangle.x1 - triangle.x3) ** 2 + (triangle.y1 - triangle.y3) ** 2
        )
        perimeter = side1 + side2 + side3
        return f"Triangle perimeter: {perimeter:.2f}"

    def visit_compound(self, compound: CompoundShape) -> str:
        total = 0.0
        for shape in compound.get_shapes():
            result = shape.accept(self)
            value = float(result.split(": ")[1])
            total += value
        return f"Compound '{compound.name}' total perimeter: {total:.2f}"


class DrawingVisitor(ShapeVisitor):
    """Visitor that generates drawing commands."""

    def visit_circle(self, circle: Circle) -> str:
        return f"Draw circle at ({circle.x}, {circle.y}) with radius {circle.radius}"

    def visit_rectangle(self, rectangle: Rectangle) -> str:
        return (
            f"Draw rectangle at ({rectangle.x}, {rectangle.y}) "
            f"with size {rectangle.width}x{rectangle.height}"
        )

    def visit_triangle(self, triangle: Triangle) -> str:
        return (
            f"Draw triangle: ({triangle.x1}, {triangle.y1}) -> "
            f"({triangle.x2}, {triangle.y2}) -> ({triangle.x3}, {triangle.y3})"
        )

    def visit_compound(self, compound: CompoundShape) -> str:
        commands = [f"Begin compound '{compound.name}'"]
        for shape in compound.get_shapes():
            commands.append(f"  {shape.accept(self)}")
        commands.append(f"End compound '{compound.name}'")
        return "\n".join(commands)


class XMLExporter(ShapeVisitor):
    """Visitor that exports shapes to XML."""

    def visit_circle(self, circle: Circle) -> str:
        return (
            f'<circle x="{circle.x}" y="{circle.y}" radius="{circle.radius}"/>'
        )

    def visit_rectangle(self, rectangle: Rectangle) -> str:
        return (
            f'<rectangle x="{rectangle.x}" y="{rectangle.y}" '
            f'width="{rectangle.width}" height="{rectangle.height}"/>'
        )

    def visit_triangle(self, triangle: Triangle) -> str:
        return (
            f'<triangle x1="{triangle.x1}" y1="{triangle.y1}" '
            f'x2="{triangle.x2}" y2="{triangle.y2}" '
            f'x3="{triangle.x3}" y3="{triangle.y3}"/>'
        )

    def visit_compound(self, compound: CompoundShape) -> str:
        lines = [f'<compound name="{compound.name}">']
        for shape in compound.get_shapes():
            lines.append(f"  {shape.accept(self)}")
        lines.append("</compound>")
        return "\n".join(lines)


class JSONExporter(ShapeVisitor):
    """Visitor that exports shapes to JSON."""

    def visit_circle(self, circle: Circle) -> str:
        return (
            f'{{"type": "circle", "x": {circle.x}, '
            f'"y": {circle.y}, "radius": {circle.radius}}}'
        )

    def visit_rectangle(self, rectangle: Rectangle) -> str:
        return (
            f'{{"type": "rectangle", "x": {rectangle.x}, '
            f'"y": {rectangle.y}, "width": {rectangle.width}, '
            f'"height": {rectangle.height}}}'
        )

    def visit_triangle(self, triangle: Triangle) -> str:
        return (
            f'{{"type": "triangle", "points": ['
            f'[{triangle.x1}, {triangle.y1}], '
            f'[{triangle.x2}, {triangle.y2}], '
            f'[{triangle.x3}, {triangle.y3}]]}}'
        )

    def visit_compound(self, compound: CompoundShape) -> str:
        shapes_json = [shape.accept(self) for shape in compound.get_shapes()]
        return (
            f'{{"type": "compound", "name": "{compound.name}", '
            f'"shapes": [{", ".join(shapes_json)}]}}'
        )


# Object Structure
class Drawing:
    """Collection of shapes that can be visited."""

    def __init__(self):
        self._shapes: list[Shape] = []

    def add(self, shape: Shape) -> None:
        self._shapes.append(shape)

    def apply(self, visitor: ShapeVisitor) -> list[str]:
        """Apply visitor to all shapes."""
        return [shape.accept(visitor) for shape in self._shapes]


def main() -> None:
    print("=" * 60)
    print("Visitor Pattern - Shape Operations Demo")
    print("=" * 60)

    # Create shapes
    circle = Circle(10, 10, 5)
    rectangle = Rectangle(0, 0, 10, 20)
    triangle = Triangle(0, 0, 10, 0, 5, 8)

    # Create compound shape
    compound = CompoundShape("House")
    compound.add(Rectangle(10, 0, 20, 15))  # Body
    compound.add(Triangle(10, 15, 30, 15, 20, 25))  # Roof
    compound.add(Circle(20, 8, 2))  # Window

    # Create drawing
    drawing = Drawing()
    drawing.add(circle)
    drawing.add(rectangle)
    drawing.add(triangle)
    drawing.add(compound)

    # Demo 1: Area calculation
    print("\n--- Area Calculator Visitor ---")
    area_calc = AreaCalculator()
    for result in drawing.apply(area_calc):
        print(f"  {result}")

    # Demo 2: Perimeter calculation
    print("\n--- Perimeter Calculator Visitor ---")
    perimeter_calc = PerimeterCalculator()
    for result in drawing.apply(perimeter_calc):
        print(f"  {result}")

    # Demo 3: Drawing commands
    print("\n--- Drawing Visitor ---")
    draw_visitor = DrawingVisitor()
    for result in drawing.apply(draw_visitor):
        print(f"  {result}")

    # Demo 4: XML export
    print("\n--- XML Export Visitor ---")
    xml_exporter = XMLExporter()
    print("<shapes>")
    for result in drawing.apply(xml_exporter):
        for line in result.split("\n"):
            print(f"  {line}")
    print("</shapes>")

    # Demo 5: JSON export
    print("\n--- JSON Export Visitor ---")
    json_exporter = JSONExporter()
    print("[")
    results = drawing.apply(json_exporter)
    for i, result in enumerate(results):
        comma = "," if i < len(results) - 1 else ""
        print(f"  {result}{comma}")
    print("]")

    # Demo 6: Single shape with multiple visitors
    print("\n--- Multiple Visitors on Single Shape ---")
    shape = Rectangle(0, 0, 15, 10)
    print("Rectangle (15x10):")
    print(f"  {shape.accept(area_calc)}")
    print(f"  {shape.accept(perimeter_calc)}")
    print(f"  {shape.accept(draw_visitor)}")

    print("\n" + "=" * 60)
    print("Benefits of Visitor Pattern:")
    print("=" * 60)
    print("1. Adds operations without modifying element classes")
    print("2. Related operations grouped in visitor class")
    print("3. Can accumulate state while visiting elements")
    print("4. Works well with Composite pattern")
    print("5. Follows Single Responsibility Principle")


if __name__ == "__main__":
    main()

File: rust/design_pattern/visitor/main_visitor.rs
// Visitor Design Pattern in Rust
// Demonstrates separating algorithms from objects they operate on

use std::f64::consts::PI;

// Visitor trait - defines operations for each element type
trait ShapeVisitor {
    fn visit_circle(&mut self, circle: &Circle);
    fn visit_rectangle(&mut self, rectangle: &Rectangle);
    fn visit_triangle(&mut self, triangle: &Triangle);
}

// Element trait - shapes that can accept visitors
trait Shape {
    fn accept(&self, visitor: &mut dyn ShapeVisitor);
    fn name(&self) -> &str;
}

// Concrete Element: Circle
struct Circle {
    radius: f64,
}

impl Circle {
    fn new(radius: f64) -> Self {
        println!("  [Circle] Created with radius: {}", radius);
        Circle { radius }
    }
}

impl Shape for Circle {
    fn accept(&self, visitor: &mut dyn ShapeVisitor) {
        println!("  [Circle] Accepting visitor...");
        visitor.visit_circle(self);
    }

    fn name(&self) -> &str {
        "Circle"
    }
}

// Concrete Element: Rectangle
struct Rectangle {
    width: f64,
    height: f64,
}

impl Rectangle {
    fn new(width: f64, height: f64) -> Self {
        println!("  [Rectangle] Created with width: {}, height: {}", width, height);
        Rectangle { width, height }
    }
}

impl Shape for Rectangle {
    fn accept(&self, visitor: &mut dyn ShapeVisitor) {
        println!("  [Rectangle] Accepting visitor...");
        visitor.visit_rectangle(self);
    }

    fn name(&self) -> &str {
        "Rectangle"
    }
}

// Concrete Element: Triangle
struct Triangle {
    base: f64,
    height: f64,
}

impl Triangle {
    fn new(base: f64, height: f64) -> Self {
        println!("  [Triangle] Created with base: {}, height: {}", base, height);
        Triangle { base, height }
    }
}

impl Shape for Triangle {
    fn accept(&self, visitor: &mut dyn ShapeVisitor) {
        println!("  [Triangle] Accepting visitor...");
        visitor.visit_triangle(self);
    }

    fn name(&self) -> &str {
        "Triangle"
    }
}

// Concrete Visitor: Area Calculator
struct AreaCalculator {
    total_area: f64,
}

impl AreaCalculator {
    fn new() -> Self {
        println!("  [AreaCalculator] Initialized with total_area: 0.0");
        AreaCalculator { total_area: 0.0 }
    }

    fn get_total_area(&self) -> f64 {
        self.total_area
    }
}

impl ShapeVisitor for AreaCalculator {
    fn visit_circle(&mut self, circle: &Circle) {
        let area = PI * circle.radius * circle.radius;
        self.total_area += area;
        println!("    -> Calculated circle area: {:.2}", area);
    }

    fn visit_rectangle(&mut self, rectangle: &Rectangle) {
        let area = rectangle.width * rectangle.height;
        self.total_area += area;
        println!("    -> Calculated rectangle area: {:.2}", area);
    }

    fn visit_triangle(&mut self, triangle: &Triangle) {
        let area = 0.5 * triangle.base * triangle.height;
        self.total_area += area;
        println!("    -> Calculated triangle area: {:.2}", area);
    }
}

// Concrete Visitor: Shape Exporter (exports to different formats)
struct ShapeExporter {
    export_data: Vec<String>,
}

impl ShapeExporter {
    fn new() -> Self {
        println!("  [ShapeExporter] Initialized with empty export buffer");
        ShapeExporter {
            export_data: Vec::new(),
        }
    }

    fn get_export(&self) -> String {
        self.export_data.join("\n")
    }
}

impl ShapeVisitor for ShapeExporter {
    fn visit_circle(&mut self, circle: &Circle) {
        let json = format!(
            r#"{{ "type": "circle", "radius": {} }}"#,
            circle.radius
        );
        self.export_data.push(json.clone());
        println!("    -> Exported circle: {}", json);
    }

    fn visit_rectangle(&mut self, rectangle: &Rectangle) {
        let json = format!(
            r#"{{ "type": "rectangle", "width": {}, "height": {} }}"#,
            rectangle.width, rectangle.height
        );
        self.export_data.push(json.clone());
        println!("    -> Exported rectangle: {}", json);
    }

    fn visit_triangle(&mut self, triangle: &Triangle) {
        let json = format!(
            r#"{{ "type": "triangle", "base": {}, "height": {} }}"#,
            triangle.base, triangle.height
        );
        self.export_data.push(json.clone());
        println!("    -> Exported triangle: {}", json);
    }
}

// Concrete Visitor: Perimeter Calculator
struct PerimeterCalculator {
    total_perimeter: f64,
}

impl PerimeterCalculator {
    fn new() -> Self {
        println!("  [PerimeterCalculator] Initialized with total_perimeter: 0.0");
        PerimeterCalculator { total_perimeter: 0.0 }
    }

    fn get_total_perimeter(&self) -> f64 {
        self.total_perimeter
    }
}

impl ShapeVisitor for PerimeterCalculator {
    fn visit_circle(&mut self, circle: &Circle) {
        let perimeter = 2.0 * PI * circle.radius;
        self.total_perimeter += perimeter;
        println!("    -> Calculated circle perimeter: {:.2}", perimeter);
    }

    fn visit_rectangle(&mut self, rectangle: &Rectangle) {
        let perimeter = 2.0 * (rectangle.width + rectangle.height);
        self.total_perimeter += perimeter;
        println!("    -> Calculated rectangle perimeter: {:.2}", perimeter);
    }

    fn visit_triangle(&mut self, triangle: &Triangle) {
        // Assuming isoceles triangle for simplicity
        let side = (triangle.height.powi(2) + (triangle.base / 2.0).powi(2)).sqrt();
        let perimeter = triangle.base + 2.0 * side;
        self.total_perimeter += perimeter;
        println!("    -> Calculated triangle perimeter: {:.2}", perimeter);
    }
}

// Object structure that holds elements
struct ShapeCollection {
    shapes: Vec<Box<dyn Shape>>,
}

impl ShapeCollection {
    fn new() -> Self {
        println!("  [ShapeCollection] Created empty collection");
        ShapeCollection { shapes: Vec::new() }
    }

    fn add(&mut self, shape: Box<dyn Shape>) {
        println!("  [ShapeCollection] Added {} to collection", shape.name());
        self.shapes.push(shape);
    }

    fn accept(&self, visitor: &mut dyn ShapeVisitor) {
        for shape in &self.shapes {
            shape.accept(visitor);
        }
    }
}

fn main() {
    println!("=== Visitor Design Pattern Demo ===\n");

    // Create shapes
    println!("1. Creating shapes:");
    let circle = Box::new(Circle::new(5.0));
    let rectangle = Box::new(Rectangle::new(4.0, 6.0));
    let triangle = Box::new(Triangle::new(3.0, 4.0));

    // Add shapes to collection
    println!("\n2. Building shape collection:");
    let mut collection = ShapeCollection::new();
    collection.add(circle);
    collection.add(rectangle);
    collection.add(triangle);

    // Use Area Calculator visitor
    println!("\n3. Applying AreaCalculator visitor:");
    let mut area_calc = AreaCalculator::new();
    collection.accept(&mut area_calc);
    println!("  [Result] Total area: {:.2}\n", area_calc.get_total_area());

    // Use Shape Exporter visitor
    println!("4. Applying ShapeExporter visitor:");
    let mut exporter = ShapeExporter::new();
    collection.accept(&mut exporter);
    println!("  [Result] Export data:");
    for line in exporter.get_export().lines() {
        println!("    {}", line);
    }

    // Use Perimeter Calculator visitor
    println!("\n5. Applying PerimeterCalculator visitor:");
    let mut perimeter_calc = PerimeterCalculator::new();
    collection.accept(&mut perimeter_calc);
    println!("  [Result] Total perimeter: {:.2}\n", perimeter_calc.get_total_perimeter());

    // Demonstrate adding new visitor without modifying shapes
    println!("6. Key Pattern Benefits:");
    println!("  - New operations added without modifying shape classes");
    println!("  - Operations on shapes are centralized in visitors");
    println!("  - Double dispatch achieved via accept/visit methods");
    println!("  - Easy to add new visitors for different algorithms");

    println!("\n=== Demo Complete ===");
}
